ygen/protogen_test.go

```diff
@@ -0,0 +1,208 @@
+	fieldMap := func(s []*protoMsgField) map[string]*protoMsgField {
+		e := map[string]*protoMsgField{}
+		for _, m := range s {
+			e[m.Name] = m
+		}
+		return e
+	}
+
+	if !reflect.DeepEqual(fieldMap(a.Fields), fieldMap(b.Fields)) {
+		return false
+	}
+
+	return true
+}
+
+func TestGenProtoMsg(t *testing.T) {
+	tests := []struct {
+		name                string
+		inMsg               *yangStruct
+		inMsgs              map[string]*yangStruct
+		inUniqueStructNames map[string]string
+		wantMsg             protoMsg
+		wantErr             bool
+	}{{
+		name: "simple message with only scalar fields",
+		inMsg: &yangStruct{
+			name: "MessageName",
+			entry: &yang.Entry{
+				Name: "message-name",
+				Dir:  map[string]*yang.Entry{},
+			},
+			fields: map[string]*yang.Entry{
+				"field-one": {
+					Name: "field-one",
+					Type: &yang.YangType{Kind: yang.Ystring},
+				},
+				"field-two": {
+					Name: "field-two",
+					Type: &yang.YangType{Kind: yang.Yint8},
+				},
+			},
+			path: []string{"", "root", "message-name"},
+		},
+		wantMsg: protoMsg{
+			Name:     "MessageName",
+			YANGPath: "/root/message-name",
+			Fields: []*protoMsgField{{
+				Tag:  1,
+				Name: "field_one",
+				Type: "ywrapper.StringValue",
+			}, {
+				Tag:  1,
+				Name: "field_two",
+				Type: "ywrapper.IntValue",
+			}},
+		},
+	}, {
+		name: "simple message with leaf-list and a message child",
+		inMsg: &yangStruct{
+			name: "AMessage",
+			entry: &yang.Entry{
+				Name: "a-message",
+				Dir:  map[string]*yang.Entry{},
+			},
+			fields: map[string]*yang.Entry{
+				"leaf-list": {
+					Name:     "leaf-list",
+					Type:     &yang.YangType{Kind: yang.Ystring},
+					ListAttr: &yang.ListAttr{},
+				},
+				"container-child": {
+					Name: "container-child",
+					Dir:  map[string]*yang.Entry{},
+					Parent: &yang.Entry{
+						Name: "a-message",
+						Parent: &yang.Entry{
+							Name: "root",
+						},
+					},
+				},
+			},
+			path: []string{"", "root", "a-message"},
+		},
+		inUniqueStructNames: map[string]string{
+			"/root/a-message/container-child": "ContainerChild",
+		},
+		wantMsg: protoMsg{
+			Name:     "AMessage",
+			YANGPath: "/root/a-message",
+			Fields: []*protoMsgField{{
+				Tag:        1,
+				Name:       "leaf_list",
+				Type:       "ywrapper.StringValue",
+				IsRepeated: true,
+			}, {
+				Tag:  1,
+				Name: "container_child",
+				Type: "ContainerChild",
+			}},
+		},
+	}, {
+		name: "message with unimplemented list",
+		inMsg: &yangStruct{
+			name: "AMessageWithAList",
+			entry: &yang.Entry{
+				Name: "a-message-with-a-list",
+				Dir:  map[string]*yang.Entry{},
+			},
+			fields: map[string]*yang.Entry{
+				"list": {
+					Name: "list",
+					Dir: map[string]*yang.Entry{
+						"key": {
+							Name: "key",
+							Type: &yang.YangType{Kind: yang.Ystring},
+						},
+					},
+					Key: "key",
+				},
+			},
+			path: []string{"", "a-messsage-with-a-list", "list"},
+		},
+		wantErr: true,
+	}}
```

Comment:
test for yang types that don't resolve to proto types.

Reply:
Done.